### PR TITLE
add leaf hub id

### PR DESCRIPTION
**Pull request by ldpliu**
https://issues.redhat.com/browse/ACM-8409

### CODE REVIEW

File: agent/pkg/status/controller/hubcluster/clusters_status_sync.go [hunk 1]
```diff
@@ -64,28 +65,36 @@ func (c *hubClusterController) Reconcile(ctx context.Context, request ctrl.Reque
...
+
+	c.syncBundle(ctx)
 
 	reqLogger.V(2).Info("Reconciliation complete.")
 
 	return ctrl.Result{}, nil
 }
 
-func (c *hubClusterController) syncBundle(ctx context.Context, route *routev1.Route) {
-	c.bundle.UpdateObject(route)
```
Comment (clyang82) [inline agent/pkg/status/controller/hubcluster/clusters_status_sync.go:87]: what is purpose to move it out of this syncBundle method?

File: operator/pkg/controllers/hubofhubs/globalhub_database.go [hunk 2]
```diff
@@ -74,6 +82,27 @@ func (r *MulticlusterGlobalHubReconciler) ReconcileDatabase(ctx context.Context,
 	return nil
 }
 
+func handleUpgrade(ctx context.Context, conn *pgx.Conn, log logr.Logger, upgradeFS embed.FS, username string) error {
```
Comment (clyang82) [inline operator/pkg/controllers/hubofhubs/globalhub_database.go:85]: I am thinking a case that if the table is altered, but the global hub agent is still old. so it can send the data without cluster_id. it reports the error which can generate lot of logs.

**Review by clyang82 — COMMENTED** [review]
Do we need to add `cluster_id` in event.local_policies table? otherwise, how to fix the problem in https://issues.redhat.com/browse/ACM-8409

**Comment by clyang82** [comment]
@ldpliu that is a big changes. Could you write down a proposal or a solution? Thanks.

**Comment by ldpliu** [comment]
@clyang82 @yanmxa 
I have changed the pr follow our discussion. Please help to take a look again.
<Will handle failed case in another pr>

File: pkg/constants/constants.go [hunk 3]
```diff
@@ -22,6 +22,8 @@ const (
 
 	DefaultAgentQPS   = float32(150)
 	DefaultAgentBurst = 300
+
+	DefaultClusterId = "00000000-0000-0000-0000-000000000000"
```
Comment (yanmxa) [inline pkg/constants/constants.go:26]: Since we using a fake ID for the existing cluster, how about generating a unique ID and make it the primary key for `the status.leaf_hub`, instead the joint primary key?

Comment (ldpliu) [reply pkg/constants/constants.go:26]: As we need to known the default id when we update the table. It is also used in dashboard part, So I think current id is necessary.

Comment (yanmxa) [reply pkg/constants/constants.go:26]: I was thinking maybe we could set an id with a value like `prefix + random` to indicate it is just a placeholder and then make the id column the primary key.
So your solution is to work in such a case. But I wonder If the ID isn't unique, It may cause other issues when we aggregate the data by ID. 

The initial purpose of introducing the `id` is because the `leafhub_name` mightn't be unique for the cluster.  But we add the `ID` also can be conflicted with each other. It seems a little bit strange.

Anyway the can work for the current scenario. We can merge this if you insist.

File: manager/pkg/statussyncer/syncers/hub_cluster_info_db_syncer.go [hunk 4]
```diff
@@ -81,47 +84,70 @@ func (syncer *hubClusterInfoDBSyncer) handleLocalObjectsBundle(ctx context.Conte
...
-
-	batchLeafhub := []models.LeafHub{}
 	for _, object := range bundle.GetObjects() {
+		clusterId := constants.DefaultClusterId
 		specificObj, ok := object.(*status.LeafHubClusterInfo)
 		if !ok {
 			continue
 		}
+		mc := clusterv1.ManagedCluster{}
+		err := (*syncer.client).Get(ctx, types.NamespacedName{
```
Comment (clyang82) [inline manager/pkg/statussyncer/syncers/hub_cluster_info_db_syncer.go:95]: I think it is better to add the clusterID from the global hub agent. We can get it from the clusterclaim CR. It is unnecessary to get the clusterclaim from the managed cluster in the global hub side. WDYT?

File: go.mod [hunk 5]
```diff
@@ -62,7 +61,10 @@ require (
 	sigs.k8s.io/yaml v1.3.0
 )
 
-require sigs.k8s.io/kube-storage-version-migrator v0.0.6-0.20230721195810-5c8923c5ff96 // indirect
+require (
+	github.com/jackc/pgconn v1.12.1 // indirect
+	sigs.k8s.io/kube-storage-version-migrator v0.0.6-0.20230721195810-5c8923c5ff96 // indirect
+)
```
Comment (clyang82) [inline go.mod:68]: unnecessary change?

File: manager/pkg/statussyncer/syncers/hub_cluster_info_syncer.go [hunk 6]
```diff
@@ -74,38 +72,58 @@ func (syncer *hubClusterInfoDBSyncer) handleLocalObjectsBundle(ctx context.Conte
...
-			Payload:     payload,
-		})
-	}
-
-	err = db.Clauses(clause.OnConflict{
-		UpdateAll: true,
-	}).CreateInBatches(batchLeafHub, 100).Error
-	if err != nil {
-		return err
+		err = db.Where("leaf_hub_name = ?", leafHubName).Find(&existingObjs).Error
```
Comment (clyang82) [inline manager/pkg/statussyncer/syncers/hub_cluster_info_syncer.go:91]: move this statement out of the for loop. it is nothing related with for loop.

File: manager/pkg/statussyncer/syncers/hub_cluster_info_syncer.go [hunk 7]
```diff
@@ -74,38 +72,58 @@ func (syncer *hubClusterInfoDBSyncer) handleLocalObjectsBundle(ctx context.Conte
...
+					Or(&models.LeafHub{
+						LeafHubName: leafHubName,
+						ClusterID:   clusterId,
+					}).
+					Updates(&models.LeafHub{
+						LeafHubName: leafHubName,
+						ClusterID:   clusterId,
+						Payload:     payload,
+					}).Error
+				if err != nil {
```
Comment (clyang82) [inline manager/pkg/statussyncer/syncers/hub_cluster_info_syncer.go:122]: can we continue to use the previous logic to update the db in batch?

**Comment by ldpliu** [comment]
> correct me if i am wrong. in your code, it is possible to insert the default clusterId `00000000-0000-0000-0000-000000000000`. since the clusterid + leaf hub name is primary key, how do you update it?

In normal case, the clusterid should not be `00000000-0000-0000-0000-000000000000`. It only used to handle upgrade case(destroyed clusters in status.leaf_hubs) or cluster claim not found case. For these cases which use the default clusterid, we use the leafhub name to identify the unique leafhub.

File: manager/pkg/statussyncer/syncers/hub_cluster_info_syncer.go (hunk 7 above)
Comment (ldpliu) [reply manager/pkg/statussyncer/syncers/hub_cluster_info_syncer.go]: > can we continue to use the previous logic to update the db in batch?

Actually, it's difficult to handle the logic use batch job.

And in the [bundle](https://github.com/stolostron/multicluster-global-hub/pull/715/files#diff-d777344f3d335e08ad16334adc3dc9471df546bc25ff446c3e1fb8d0a106ec94R76), it only have one item, So I have changed the logic and it only update db once also.

**Comment by yanmxa** [comment]
Reference: The deletion https://github.com/stolostron/multicluster-global-hub/blob/main/manager/pkg/specsyncer/spec2db/controller/managedhub_spec_sync.go

File: manager/pkg/statussyncer/syncers/hub_cluster_info_syncer.go [hunk 8]
```diff
@@ -74,38 +72,58 @@ func (syncer *hubClusterInfoDBSyncer) handleLocalObjectsBundle(ctx context.Conte
...
+			err := db.Create(&models.LeafHub{
+				LeafHubName: leafHubName,
+				ClusterID:   clusterId,
+				Payload:     payload,
+			}).Error
+			return err
+		}
+		err = db.Model(&models.LeafHub{}).
+			Where(&models.LeafHub{
+				LeafHubName: leafHubName,
```
Comment (clyang82) [inline manager/pkg/statussyncer/syncers/hub_cluster_info_syncer.go:117]: Add `deleted_at` is NULL as a condition.

**Comment by yanmxa** [comment]
/lgtm 

once it is merged, I need to consolidate by this: https://github.com/stolostron/multicluster-global-hub/pull/728
